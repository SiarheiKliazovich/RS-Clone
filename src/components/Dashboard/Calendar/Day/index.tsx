import "./day.scss";
import { FunctionComponent } from "react";
import { CalendarDayType } from "../../../../types";
import classNames from "classnames";
import moment from "moment";

const Day: FunctionComponent<CalendarDayType> = ({
  day,
  status,
}: CalendarDayType) => {
  const data = [
    { header: "Non paid", value: status.nonPaid.length, style: null },
    { header: "Paid", value: status.paid.length, style: null },
    { header: "In progress", value: status.inProgress.length, style: null },
    { header: "Declined", value: status.declined.length, style: null },
    { header: "Finished", value: status.finished.length, style: null },
    { header: "Deadline", value: status.deadline.length, style: null },
  ];

  const markers = data
    .filter((item) => item.value !== 0)
    .map((item) => (
      <li className="day__item" key={item.header}>
        <span>{item.header}</span>
        <span>{item.value}</span>
      </li>
    ));

  return (
    <div
      className={classNames("day", {
        "day--today": moment().isSame(day, "day"),
        "day--current": moment().isSame(day, "month"),
      })}
    >
      <h6>{day.format("DD-MMM")}</h6>
      <ul className="day__list">{markers}</ul>
    </div>
  );
};

export default Day;
