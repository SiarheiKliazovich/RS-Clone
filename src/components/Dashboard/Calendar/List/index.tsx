import "./list.scss";
import moment from "moment";
import { FunctionComponent, ReactElement } from "react";
import { CalendarListType } from "../../../../types";
import { IOrder } from "../../../../interfaces/order";
import classNames from "classnames";
import { useTranslation } from "react-i18next";

const Index: FunctionComponent<CalendarListType> = ({
  day,
  status,
}: CalendarListType) => {
  const { t } = useTranslation("dataLang");

  const elements: ReactElement[] = [];

  const setMarkerText = (status: string): { text: string; style: string } => {
    switch (status) {
      case "NON-PAID":
        return { text: t("select.nonPaid"), style: "marker--non-pain" };
      case "PAID":
        return { text: t("select.paid"), style: "marker--pain" };
      case "DECLINED":
        return { text: t("select.declined"), style: "marker--declined" };
      case "IN PROGRESS":
        return { text: t("select.inProgress"), style: "marker--in-progress" };
      case "FINISHED":
        return { text: t("select.finished"), style: "marker--finished" };
      case "DEADLINE":
        return { text: t("select.deadline"), style: "marker--deadline" };
      default:
        return { text: t("select.deadline"), style: "marker--deadline" };
    }
  };

  for (const key in status) {
    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
    // @ts-ignore
    status[key].forEach((item: IOrder) => {
      const s = moment(day).isSame(item.deadlineDate, "day")
        ? "DEADLINE"
        : item.status;

      const { style, text } = setMarkerText(s);

      const el = (
        <li
          className={classNames("list__item", {
            "list__item--warning": moment(day).isSame(item.deadlineDate, "day"),
          })}
          key={item._id}
        >
          <span className="list__column">
            #{item._id.slice(-9).toUpperCase()}
          </span>
          <span className="list__column">
            <img className="list__image" src={item.image} alt="" />
          </span>
          <span className="list__column">
            {moment(item.date).format("D MMMM")}
          </span>
          <span className="list__column">{item.price}</span>
          <span className="list__column">
            <span className={classNames("marker", style)}>{text}</span>
          </span>
          <span
            className={classNames("list__column list__column--last", {
              "list__column--warning": day.isSame(
                moment(item.deadlineDate),
                "day"
              ),
            })}
          >
            {moment(item.deadlineDate).format("D MMMM")}
          </span>
        </li>
      );
      elements.push(el);
    });
  }

  return (
    <div className="list">
      <h4 className="list__header">{moment(day).format("D MMMM")}</h4>
      <ul className="list__list">
        <li className="list__item">
          <span className="list__column">{t("calendar.list.id")}</span>
          <span className="list__column">{t("calendar.list.product")}</span>
          <span className="list__column">{t("calendar.list.date")}</span>
          <span className="list__column">{t("calendar.list.price")}</span>
          <span className="list__column">{t("calendar.list.status")}</span>
          <span className="list__column list__column--last">
            {t("calendar.list.deadline")}
          </span>
        </li>
        {elements}
      </ul>
    </div>
  );
};

export default Index;
