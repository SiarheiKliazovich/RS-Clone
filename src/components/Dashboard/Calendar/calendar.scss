.calendar {

}

.calendar__container {
  width: 700px;
  display: flex;
  flex-wrap: wrap;
}

.calendar__title {
  width: 100px;
  display: block;
  text-align: center;
}

.calendar__day {
  width: 100px;
  height: 40px;
  display: block;
  border: 1px solid yellowgreen;

  &--today {
    color: red;
  }
}
