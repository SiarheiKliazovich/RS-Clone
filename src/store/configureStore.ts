/* eslint-disable @typescript-eslint/ban-ts-comment */
import { applyMiddleware, createStore } from "redux";
import thunkMiddleware from "redux-thunk";

import loggerMiddleware from "./logger";
import rootReducer from "../reducers";

const configureStore = () => {
  const middlewares = [loggerMiddleware, thunkMiddleware];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const store = createStore(rootReducer, middlewareEnhancer);

  // @ts-ignore:next-line
  if (process.env.NODE_ENV !== "production" && module.hot) {
    // @ts-ignore:next-line
    module.hot.accept("./../reducers", () => store.replaceReducer(rootReducer));
  }

  return store;
};

export default configureStore;
