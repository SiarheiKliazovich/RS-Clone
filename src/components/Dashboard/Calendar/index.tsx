import "./calendar.scss";
import { FunctionComponent, useEffect, useState } from "react";
import moment, { Moment } from "moment";
import "moment/locale/ru";
import Day from "./Day";
import List from "./List";
import { CalendarDayType } from "../../../types";
import { getOrdersBySellerId } from "../../../services/apiOrders";
import { useSelector } from "react-redux";
import { IStore } from "../../../interfaces/store";
import { IUser } from "../../../interfaces/user";
import { IOrder } from "../../../interfaces/order";
import { useTranslation } from "react-i18next";

const Calendar: FunctionComponent = () => {
  const language = useSelector((state: IStore) => state.langReducer.lang);
  moment.locale(language);
  const { t } = useTranslation("dataLang");

  const user: IUser = useSelector((state: IStore) => state.auth.user);

  const [orders, setOrders] = useState<IOrder[]>([]);
  const [ordersDeadline, setOrdersDeadline] = useState<IOrder[]>([]);
  const [focusedDay, setFocusedDay] = useState({ day: null, status: null });
  const [viewedMonth, setViewedMonth] = useState<number>(moment().month());

  const deadlineStep = 7;
  const dayDataArr: CalendarDayType[] = [];

  useEffect(() => {
    getOrdersBySellerId(user._id)
      .then((orders: IOrder[]) => {
        setOrdersDeadline(prepareOrdersToDeadline(orders));
        setOrders(orders);
      })
      .catch((e) => console.log(e));

    const prepareOrdersToDeadline = (orders: IOrder[]) => {
      return orders
        .filter((order: IOrder) => order.status !== "FINISHED")
        .filter((order: IOrder) => order.status !== "DECLINED")
        .map((order: IOrder) => {
          order.deadlineDate = moment(order.date)
            .subtract(-deadlineStep, "day")
            .toISOString();
          return order;
        });
    };
  }, []);

  const createCalendar = () => {
    const size = 42; // 7 days * 6 weeks
    let start =
      moment()
        .locale("en") // important!
        .subtract(moment().month() - viewedMonth, "month")
        .startOf("month")
        .startOf("isoWeek")
        .dayOfYear() - 1;

    if (language !== "ru") start = start - 1;

    const end = start + size;
    const now = moment().dayOfYear();

    const startI = now - end;
    const endI = now - start;

    for (let i = startI; i < endI; i++) {
      const day = moment().subtract(i, "day");
      const dayOrders = orders.filter((order) => {
        return day.year() !== moment().year()
          ? day.format("DD-MM") === moment(order.date).format("DD-MM")
          : day.isSame(order.date, "day");
      });

      addDataToDay(day, dayOrders);
    }
  };

  const addDataToDay = (day: Moment, orders: IOrder[]) => {
    const status = {
      nonPaid: [] as IOrder[],
      paid: [] as IOrder[],
      inProgress: [] as IOrder[],
      declined: [] as IOrder[],
      finished: [] as IOrder[],
      deadline: [] as IOrder[],
    };

    orders.forEach((order: IOrder) => {
      switch (order.status) {
        case "NON-PAID":
          status.nonPaid.push(order);
          break;
        case "PAID":
          status.paid.push(order);
          break;
        case "IN PROGRESS":
          status.inProgress.push(order);
          break;
        case "DECLINED":
          status.declined.push(order);
          break;
        case "FINISHED":
          status.finished.push(order);
          break;
      }
    });

    const ordersWithDeadline = ordersDeadline.filter((order) =>
      moment(order.deadlineDate).isSame(day, "day")
    );

    if (ordersWithDeadline.length) {
      status.deadline.push(...ordersWithDeadline);
    }

    dayDataArr.push({
      day: day,
      isActiveMonth: day.month() === viewedMonth,
      status: status,
    });
  };

  createCalendar();

  const changeMonth = (direction: number) => {
    let value = viewedMonth + direction;

    if (viewedMonth === 11 && direction === 1) value = 0;
    if (viewedMonth === 0 && direction === -1) value = 11;

    setViewedMonth(value);
  };

  const titlesElements = Array(7)
    .fill(0)
    .map((item: string, i: number) => (
      <span className="calendar__title" key={i}>
        {moment().startOf("week").subtract(-i, "day").format("dddd")}
      </span>
    ));

  const daysComponents = dayDataArr
    .reverse()
    .map((item: CalendarDayType) => (
      <Day
        day={item.day}
        status={item.status}
        isActiveMonth={item.isActiveMonth}
        key={item.day.format("DD-MM")}
        setFocusedDay={setFocusedDay}
      />
    ));

  return (
    <div className="calendar">
      <h1 className="calendar__header">{t("calendar.header")}</h1>
      <div className="calendar__wrapper">
        <div className="calendar__head">
          <div>
            <button onClick={() => changeMonth(-1)}>
              {t("calendar.buttons.previous")}
            </button>
            <button onClick={() => changeMonth(1)}>
              {t("calendar.buttons.next")}
            </button>
          </div>
          <h2 className="calendar__subheader">
            {moment()
              .subtract(1 - viewedMonth, "month")
              .format("MMMM")}
          </h2>
          <div>
            <button onClick={() => setViewedMonth(moment().month())}>
              {t("calendar.buttons.current")}
            </button>
          </div>
        </div>
        <div className="calendar__container">{titlesElements}</div>
        <div className="calendar__container">{daysComponents}</div>
      </div>
      {focusedDay.day ? (
        <List day={focusedDay.day} status={focusedDay.status} />
      ) : null}
    </div>
  );
};

export default Calendar;
