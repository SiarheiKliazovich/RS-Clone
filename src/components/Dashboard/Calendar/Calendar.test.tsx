import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import Calendar from ".";

const mockStore = configureStore([]);

describe("Calendar", () => {
  // eslint-disable-next-line
  let store: any;

  beforeEach(() => {
    store = mockStore({
      appInterface: {
        accentColor: {
          static: "#000",
          hover: "#fff",
        },
      },
      langReducer: {
        lang: "en",
      },
      auth: {
        user: {
          role: "USER",
        },
      },
    });
  });

  test("renders correctly", () => {
    const { container } = render(
      <Provider store={store}>
        <Calendar />
      </Provider>
    );
    expect(container).toMatchSnapshot();
  });
});
