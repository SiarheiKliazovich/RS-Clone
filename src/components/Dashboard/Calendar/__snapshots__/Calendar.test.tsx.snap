// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar renders correctly 1`] = `
<div>
  <div
    class="calendar"
  >
    <div
      class="calendar__wrapper"
    >
      <div
        class="calendar__head"
      >
        <div
          class="calendar__head-column"
        >
          <button
            class="button calendar__button--left"
            style="background-color: rgb(0, 0, 0);"
          >
            calendar.buttons.previous
          </button>
          <button
            class="button calendar__button--right"
            style="background-color: rgb(0, 0, 0);"
          >
            calendar.buttons.next
          </button>
        </div>
        <h2
          class="calendar__subheader"
        >
          February
        </h2>
        <div
          class="calendar__head-column"
        >
          <button
            class="button"
            style="background-color: rgb(0, 0, 0);"
          >
            calendar.buttons.current
          </button>
        </div>
      </div>
      <div
        class="calendar__container"
      >
        <span
          class="calendar__title"
        >
          Sun
        </span>
        <span
          class="calendar__title"
        >
          Mon
        </span>
        <span
          class="calendar__title"
        >
          Tue
        </span>
        <span
          class="calendar__title"
        >
          Wed
        </span>
        <span
          class="calendar__title"
        >
          Thu
        </span>
        <span
          class="calendar__title"
        >
          Fri
        </span>
        <span
          class="calendar__title"
        >
          Sat
        </span>
      </div>
      <div
        class="calendar__container"
      >
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            29
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            30
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            31
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            01
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            02
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            03
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            04
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            05
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            06
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            07
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            08
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            09
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            10
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            11
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            12
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            13
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            14
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            15
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            16
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            17
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            18
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            19
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            20
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            21
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            22
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            23
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day day--today"
        >
          <h6
            class="day__header day__header--active"
          >
            24
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            25
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            26
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            27
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header day__header--active"
          >
            28
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            01
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            02
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            03
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            04
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            05
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            06
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            07
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            08
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            09
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            10
          </h6>
          <ul
            class="day__list"
          />
        </div>
        <div
          class="day"
        >
          <h6
            class="day__header"
          >
            11
          </h6>
          <ul
            class="day__list"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
